### ISSUE TITLE
Odd assertion error since v1.30

### ISSUE THREAD

**Issue by ping-localhost** [grpc/grpc#24469]
### What version of gRPC and what language are you using?
This bug occurs in v1.30 and up

### What operating system (Linux, Windows,...) and version?
- MacOS Catalina 10.15.7 (local machine running client)
- Ubuntu bionic (when project is deployed using v1.30)

### What runtime / compiler are you using (e.g. python version or version of gcc)
Unrelated.

### What did you do?
I'm having a pretty odd issue with PECL GRPC v1.30 and up (but not v1.29.1 and below) and I would love to get some input or pointers. Since v1.30 I'm getting an assert error (specified below). I brought this up in [Gitter](https://gitter.im/grpc/grpc?at=5f88665b955f6d78a92ebb63) and someone told me to make a bug report, so here it is.

This assert is thrown by the following line: https://github.com/grpc/grpc/blob/master/src/core/tsi/ssl_transport_security.cc#L473

This issue is pretty complex, so I haven't build a reproduction yet, but I'll commit some time to it if it's really necessary. I'm hoping that the data I've provided below is clear enough.

### What did you expect to see?
I expected the call to finish, as it has before, but it doesn't.

### What did you see instead?
```
[PHP-FPM    ] Oct 15 13:42:48 |INFO   | REQUES Matched route "app_overview_list". method="GET" request_uri="https://127.0.0.1:8000/api/account/personal/list?limit=500&offset=0" route="app_overview_list" route_parameters={"_controller":"App\\Controller\\OverviewController::list","_route":"app_overview_list","entity":"personal"}
[PHP-FPM    ] Oct 15 13:42:48 |INFO   | SECURI Guard authentication successful! authenticator="App\\Security\\StubAuthenticator" token={"Symfony\\Component\\Security\\Guard\\Token\\PostAuthenticationGuardToken":"PostAuthenticationGuardToken(user=\"Foo Bar\", authenticated=true, roles=\"ROLE_FOOBAR\")"}
[PHP-FPM    ] E1015 15:42:48.435412000 4602330560 ssl_transport_security.cc:473]     assertion failed: (int)peer->property_count == current_insert_index
[Web Server ] Oct 15 15:42:48 |INFO   | SERVER GET  (200) /favicon.ico ip="127.0.0.1"
[Web Server ] Oct 15 15:42:48 |ERROR  | SERVER issue with server callback error="unable to fetch the response from the backend: unexpected EOF"
```

### Anything else we should know about your project / environment?
On the client side (PHP) I use the following to connect to the server (through NGINX with a GRPC pass):
```php
return new AccountAPIClient($host, [
    'credentials' => Grpc\ChannelCredentials::createSsl(\file_get_contents($caFilePath)),
]);
```

On the server side (Golang) we use the following:
```golang
// Set TLS options
grpc.Creds(credentials.NewTLS(&tls.Config{
    ClientAuth:   tls.RequireAndVerifyClientCert,
    Certificates: []tls.Certificate{certificate},
    ClientCAs:    certPool,
})),
```

The NGINX config we use (which creates the actual mTLS connection):
```nginx
server {
  listen 7790 ssl http2;
  client_max_body_size 15m;
  server_name foo.bar.com;

  ssl_certificate /etc/freeipa/certs/server.crt;
  ssl_certificate_key /etc/freeipa/certs/server.key; 

  location ~ ^/foo.bar.* {
    grpc_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
    grpc_ssl_certificate /etc/private/ssl/foobar/foobar-client.crt;
    grpc_ssl_certificate_key /etc/private/ssl/foobar/foobar-client.key;
    grpc_pass grpcs://foobarservers;
    access_log /var/log/nginx/access_foobar_7790.log;
    error_log /var/log/nginx/error_foobar_7790.log;
  }
}
```

**Comment by stanley-cheung** [comment]
So I suppose this can only be reproduced when PHP is invoked via Nginx? Yes a reproducible test case would be great. 

But let me get this straight. Which of the following is your setup?

1. some client makes a request to Nginx, possibly for a .php page, PHP-FPM is dispatched to handle the PHP request. There you create a PHP gRPC client, and makes a connection to a Go gRPC server.

2. a PHP client is making a gRPC request to Nginx. Nginx is re-routing that grpc request to a Go gRPC server with `grpc_pass`?

It isn't clear to me where PHP and Nginx are sitting in the end-to-end path from your description. Can you clarify that please? Thanks.

**Comment by ping-localhost** [comment]
@stanley-cheung It's the second setup. A PHP client makes a gRPC request to NGINX and which is then rerouted. When using haproxy the same issue occurs, so it really looks to be more a PHP client side issue than a `grpc_pass` issue. 🤔

If this doesn't add enough information, I'll try and build something reproducible tomorrow :)

**Comment by stanley-cheung** [comment]
I see. So the set up is

```
PHP gRPC Client <--> Nginx <--> Go gRPC Server
```

In that case, does this error happen if you remove Nginx from the picture (just for the sake of understanding the problem a little better?) i.e. what if you connect the PHP gRPC Client directly to a <lang> gRPC server directly?

**Comment by stanley-cheung** [comment]
And one more question, where does PHP-FPM come into play? Asking because I see PHP-FPM as the process spitting out that log line error of `ssl_transport_security.cc:473` in the original post. Does this mean that there's another server serving some front-end requests and that server is running PHP-FPM on the side to execute the .php scripts? Just trying to get to your set up as close as possible. If that part can be eliminated, i.e. if the problem can be reproduced with the PHP gRPC client running from the CLI that would simplify things. However, if the problem _has_ to be reproduced with this extra PHP-FPM process, then it introduces another layer of complexity.

**Comment by ping-localhost** [comment]
> In that case, does this error happen if you remove Nginx from the picture (just for the sake of understanding the problem a little better?) i.e. what if you connect the PHP gRPC Client directly to a <lang> gRPC server directly?

I'll try that and report back @stanley-cheung, but the issue also happens with haproxy, so I'm pretty certain that it's not NGINX, but def. better to test it and be sure. 👍 

---

The PHP-FPM error you see if because of me running `symfony serve` to run the client locally. Sorry for the confusion.

**Comment by ping-localhost** [comment]
@stanley-cheung I've been able to setup a testing repository now, but there I'm not getting any issues even though it should be an exact replication of what I have. After some checking it seems that the freeipa SSL certificates (which NGINX is using) are causing issues, not the actual GRPC server.

This is the certificate I'm currently using (anonymized):
```
Certificate:
    Data:
        Version: 3 (0x2)
        Serial Number: 361 (0x169)
        Signature Algorithm: sha256WithRSAEncryption
        Issuer: O = I.FOOBAR.NL, CN = Production CA
        Validity
            Not Before: Jun  12 05:30:34 2020 GMT
            Not After : Jun  12 05:30:34 2021 GMT
        Subject: O = I.FOOBAR.NL, CN = sta-hello-world-in.stage.virt.i.foobar.nl
        Subject Public Key Info:
            Public Key Algorithm: rsaEncryption
                RSA Public-Key: (2048 bit)
                Modulus:
                    REMOVED
                Exponent: 65537 (0x10001)
        X509v3 extensions:
            X509v3 Authority Key Identifier: 
                keyid:REMOVED

            Authority Information Access: 
                OCSP - URI:http://ipa-ca.i.foobar.nl/ca/ocsp

            X509v3 Key Usage: critical
                Digital Signature, Non Repudiation, Key Encipherment, Data Encipherment
            X509v3 Extended Key Usage: 
                TLS Web Server Authentication, TLS Web Client Authentication
            X509v3 CRL Distribution Points: 

                Full Name:
                  URI:http://ipa-ca.i.foobar.nl/ipa/crl/MasterCRL.bin
                CRL Issuer:
                  DirName:O = ipaca, CN = Certificate Authority

            X509v3 Subject Key Identifier: 
                REMOVED
            X509v3 Subject Alternative Name: 
                DNS:hello-world-staging.i.foobar.nl, othername:<unsupported>, othername:<unsupported>, DNS:sta-hello-world-in.stage.virt.i.foobar.nl
```

This is the cert that works whilst testing (anonymized in the same way):
```
Certificate:
    Data:
        Version: 3 (0x2)
        Serial Number:
            REMOVED
    Signature Algorithm: sha256WithRSAEncryption
        Issuer: O=mkcert development CA, OU= 37925797+ping-localhost@users.noreply.github.com (Mitchell), CN=mkcert mitchell@mitchell.i.bitonic.nl (Mitchell)
        Validity
            Not Before: Jun  1 00:00:00 2019 GMT
            Not After : Oct 20 11:10:59 2030 GMT
        Subject: O=mkcert development certificate, OU= 37925797+ping-localhost@users.noreply.github.com (Mitchell)
        Subject Public Key Info:
            Public Key Algorithm: rsaEncryption
                Public-Key: (2048 bit)
                Modulus:
                    REMOVED
                Exponent: 65537 (0x10001)
        X509v3 extensions:
            X509v3 Key Usage: critical
                Digital Signature, Key Encipherment
            X509v3 Extended Key Usage: 
                TLS Web Server Authentication
            X509v3 Basic Constraints: critical
                CA:FALSE
            X509v3 Authority Key Identifier: 
                keyid:REMOVED

            X509v3 Subject Alternative Name: 
                DNS:example.com, DNS:*.example.com, DNS:example.test, DNS:localhost, IP Address:127.0.0.1, IP Address:0:0:0:0:0:0:0:1
    Signature Algorithm: sha256WithRSAEncryption
         REMOVED
```

Anyone able to give any insight in why my currently live certificate (that works just fine before v1.30) doesn't work now. It's seems to be related to this MR: https://github.com/grpc/grpc/pull/22863

**Comment by stanley-cheung** [comment]
Would this be reproducible just with a simple client that can be launched with `symfony serve`? Does Nginx have to be run to cause the error to happen?

**Comment by ping-localhost** [comment]
> Would this be reproducible just with a simple client that can be launched with `symfony serve`? Does Nginx have to be run to cause the error to happen?

Good question @stanley-cheung. I've just checked and NGINX does not have to run for this error to happen. When using this certificate in the Golang server the same happens if I directly connect to it from the PHP client (either by using `php ./client/index.php` or by running `symfony serve`). 

So it seems that the cert that I'm using is hard to parse by `grpc/grpc` for some reason 🤔

**Comment by ping-localhost** [comment]
> Would this be reproducible just with a simple client that can be launched with `symfony serve`? Does Nginx have to be run to cause the error to happen?

Good question @stanley-cheung. I've just checked and NGINX does not have to run for this error to happen.

When using the "_invalid_" certificate in the Golang server the same happens if I directly connect to it from the PHP client (either by using `php ./client/index.php` or by running `symfony serve`). 

So it seems that the certificate is not being parsed correctly by `grpc/grpc` somehow. 🤔 

---

[I've published the simple setup I'm using locally to test](https://github.com/ping-localhost/grpc-bug-report). I cannot share the messed up certificate though as that leaks a lot of information, sorry.

**Comment by ZhenLian** [comment]
Ah, I think the problem is caused by https://github.com/grpc/grpc/blob/master/src/core/tsi/ssl_transport_security.cc#L356-L389, where `otherName` is currently not parsed as peer information. The fix should be straightforward: we either increase the index even when a not recognized SAN name passed in(preferred), or we remove that ASSERT check.
Thanks @stanley-cheung for working with the user on this issue! I wouldn't be able to spot the problem without the initial investigation work. I will work a fix in the next couple of days.

**Comment by stanley-cheung** [comment]
@ZhenLian Thanks for looking into this!

**Comment by ping-localhost** [comment]
Thank you both for taking the time to look into it at all. 😊

**Comment by ping-localhost** [comment]
I hereby confirm that the fix works. Thanks again @stanley-cheung and @ZhenLian

**Comment by ZhenLian** [comment]
Oh, I just saw this update :) Great to hear it works for you!
IIRC, gRPC releases new minor versions every month, but I am not sure about the patch version. Looking at https://github.com/grpc/grpc/releases, it seems the patch is weekly released or bi-weekly released. Maybe @stanley-cheung can confirm this?

**Comment by ping-localhost** [comment]
@stanley-cheung @ZhenLian Last release was 21 days ago. Can anyone release this fix please/tell me when it will be released? 😅

**Comment by ZhenLian** [comment]
@ping-localhost You can find the release information at https://github.com/grpc/grpc/issues/24589#issuecomment-730317941. Please let me know if you have any other questions. Thanks!
